### DISCUSSION TITLE
ChatGPT (and other models) discuss PRADO & making PRADO better...

### DISCUSSION THREAD

**Discussion by belisoful** [pradosoft/prado#870] [General]
Namaste!

If you haven't played with ChatGPT, or the Bing/google spin-offs, I suggest you start playing with it.  It is "the future."

I asked it to make my Continued Fraction function better, and yup.  It did a better job doing code than me in some instances.  There are issues with it though; (it'll only get better).  It was difficult to tweak the code into a better state, so I had to do the coding by hand, ultimately.  It is an amazing tool.  It found several weak points in my algorithm in seconds that I had been noodling/delaying on for weeks (with so much other code to work on).

I've been asking it about its familiarity with PRADO.  It knows our project.    Awesome.  Let's ask it some questions.

ChatGPT has identified Asset (File/Directory manipulation) and Asset publishing as a weak point.  I am addressing this right now, actually.  And not in a way that an AI is capable of putting together.  at least today. (the new asset manager it very simple and yet the processes are broken down into pieces that can be used in a complex way; the solutions are very innovative [in my sense])

It also identified weak documentation and a lack of examples as an issue for PRADO.  Which I agree with.  I probably need to write it, but the official documentation should probably have pages for Composer-PRADO extensions, dynamic events, fx Events, behaviors (and behavior manager with specific behaviors, like TimeZone), cron, and permissions pages plus a blurb on "TWebDecorator" in the "skin" page. Those are key advancements for PRADO, in my estimation.  sigh, documentation.

I asked it to describe how TMap works, and it did.  Then I asked it to rewrite the header documentation.  Here is what it said:

```
/**
 * TMap class represents a collection of key-value pairs where keys and values
 * can be of any PHP type. The keys are unique, and their order is not fixed.
 * TMap implements the IteratorAggregate interface and can be iterated over
 * using a foreach loop. It also implements the Countable interface, allowing
 * you to determine the number of key-value pairs in the TMap.
 *
 * TMap can be used in various situations where you need to store and access data
 * using a key-value pair structure. For example, you can use TMap to store
 * configuration settings for your application or to store information about
 * users in a database.
 *
 * You can add and remove key-value pairs from a TMap, retrieve values by key,
 * and perform other operations such as merging TMaps or converting a TMap to
 * an array.
 *
 * @package System.Collections
 * @since 3.1.0
 * @version $Id$
 */
```
The TMap header PHP doc currently states this:

```
/**
 * TMap class
 *
 * TMap implements a collection that takes key-value pairs.
 *
 * You can access, add or remove an item with a key by using
 * {@link itemAt}, {@link add}, and {@link remove}.
 * To get the number of the items in the map, use {@link getCount}.
 * TMap can also be used like a regular array as follows,
 * <code>
 * $map[$key]=$value; // add a key-value pair
 * unset($map[$key]); // remove the value with the specified key
 * if(isset($map[$key])) // if the map contains the key
 * foreach($map as $key=>$value) // traverse the items in the map
 * $n=count($map);  // returns the number of items in the map
 * </code>
 *
 * @author Qiang Xue <qiang.xue@gmail.com>
 * @since 3.0
 * @method void dyAddItem(mixed $key, mixed $value)
 * @method void dyRemoveItem(mixed $key, mixed $value)
 * @method mixed dyNoItem(mixed $returnValue, mixed $key)
 */
```
Some of the AI description is better than existing, though I do like the existing documentation.

My point is that these language models may actually help make better documentation for us.  This can be applied to every class and every function to actually help create better documentation.  Obviously, subject to human review, but this may be a powerful tool to update and better the PRADO Php Doc.

ChatGPT identified that image processing was weak in PRADO.  This is also getting [somewhat] addressed in the new asset publishing code.  The strange thing is that it found "System.Drawing" and it says that there is a TGDImage and a TGDThumbnail class in there.  it seems to be bringing in some .NET things into the mix on further investigation.  it apologized for hallucinating and gave me an over simplified (but working) implementation of TGDImage.

What is your experience with ChatGPT in relation to PRADO?

**Comment by ctrlaltca** [comment]
I played a bit with it, starting from the obvious things like making it go mad and messing it up its answers.
I also had it code a few small programs, and convert code from a language to another.
That's quite impressive, even if it makes some basic mistakes, eg. like assuming that a python library as an exact match in other languages. Still, a huge time saver.
The most impressive thing to me is its ability to express itself in a human-comprehensible way, with all the possible variations (be more formal / happy, talk like a 10 years old, etc..).
I'm quite impressed by the documentation example. The existing documentation assumes you already know what a TMap is used for, just exposing the most common method names and an example (in case you forgot the exact method name or how to use it).
The AI-generated doc is more a gentle introduction to what it is and how it works. You'll probably only need to read it once in your life to grasp the basic concept and after that you'll be better suited with the old doc, but it's still wow.

**Comment by belisoful** [comment]
I wanted to try something simple to start, like TMap.

It gives reasonable php doc for more complex classes like TCronModule though is skips out on specifics that should be mentioned.  Those specifics could be included in the prompt.  I asked it for the PHP Doc for TPermissionsManager and it did not include RBAC as part of the description, though when asked, it said that it was an RBAC implementation.

Tailoring the prompt to include specific elements may be the way to go to ensure all the details are there.

I do like the idea of reducing the linguistic barrier in understanding the platform for less experienced programmers.
